File: windows/AirBridge.Core/Interfaces/ITransferSession.cs
namespace AirBridge.Core.Interfaces;

/// <summary>State of a file transfer session.</summary>
public enum TransferState { Pending, Active, Paused, Completed, Failed, Cancelled }

/// <summary>
/// Represents a single file transfer session (send or receive).
/// Progress is reported via <see cref="ProgressChanged"/>.
/// </summary>
public interface ITransferSession : IDisposable
{
    string SessionId { get; }
    string FileName { get; }
    long TotalBytes { get; }
    long TransferredBytes { get; }
    TransferState State { get; }
    bool IsSender { get; }

    Task StartAsync(CancellationToken cancellationToken = default);
    Task PauseAsync();
    Task ResumeAsync(CancellationToken cancellationToken = default);
    Task CancelAsync();

    event EventHandler<long> ProgressChanged;     // arg: bytes transferred so far
    event EventHandler<TransferState> StateChanged;
}

File: android/app/src/main/java/com/airbridge/app/core/interfaces/ITransferSession.kt
package com.airbridge.app.core.interfaces

import kotlinx.coroutines.flow.Flow

enum class TransferState { PENDING, ACTIVE, PAUSED, COMPLETED, FAILED, CANCELLED }

/**
 * Represents a single file transfer session (send or receive).
 * [progressFlow] emits bytes-transferred updates for UI binding.
 */
interface ITransferSession {
    val sessionId: String
    val fileName: String
    val totalBytes: Long
    val isSender: Boolean
    val stateFlow: Flow<TransferState>
    val progressFlow: Flow<Long>  // bytes transferred so far

    suspend fun start()
    suspend fun pause()
    suspend fun resume()
    suspend fun cancel()
}
